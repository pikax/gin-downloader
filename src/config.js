/**
 * Created by rodriguesc on 03/03/2017.
 */


