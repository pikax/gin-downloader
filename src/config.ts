/**
 * Created by pikax on 16/07/2017.
 */
import {RequestStrategy, Options, CoreOptions, strategies} from "./request/index";

// import {RequestRetryStrategy} from "./request/requestRetryStrategy";
// import {RequestRetryStrategy} from "./request/requestRetryStrategy";



export interface GinPoolConfig {
  simultaneousRequests?: number;
  requestInterval?: number;

  match: RegExp;
}


//
// export interface GinSiteConfig {
//   strategy: RequestStrategy;
// }

export interface GinConfig {
  maxRetries?: number;
  timeout?: number;
  interval?: number;


  pooling?: {[poolId: string]: GinPoolConfig};

  sites?: {[key: string]: RequestStrategy};

  request?: CoreOptions;
}

/*TODO this shouldn't be a function, this should be a const only, but the default config is called by the strategies
* that means when we load the strategies module, it needs to load this one first but this uses strategies.
* */
const DefaultConfig: () => GinConfig = () => ({
  maxRetries: 50,
  timeout: 10000,
  interval: 1000,

  pooling: {
    MangafoxSearch: {
        requestInterval: 5000,
        match: /mangafox\.me\/search\.php/
    }
  },


  sites: {
    batoto: strategies.retry,
    mangafox: strategies.retry,
    mangapanda: strategies.retry,
    mangahere: strategies.retry,
    kissmanga: strategies.cloudflare,
  },

  request: {
    jar: true,
    gzip: true,
    followAllRedirects: true,
    forever: true,
    headers: {
      "Accept-Charset": "utf-8;q=0.7,*;q=0.3",
      "Accept-Language": "en-US,en;q=0.8",
      "Connection": "keep-alive",
      "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8",
      "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/47.0.2526.111 Safari/537.36",
      "Accept-Encoding": "gzip,deflate",
    }
  }
});

export class Config {
  private _use: GinConfig;
  private _config: GinConfig;


  get defaultConfig() { return DefaultConfig(); }

  get config(): GinConfig { return this._config || (this._config = this.buildConfig()); }


  get use(): GinConfig {
    return this._use;
  }
  set use(use: GinConfig) {
    this._use = use;

    this._config = this.buildConfig(use);
  }

  reset() {
    this._config = this.defaultConfig;
    this._use = null;
  }

  private buildConfig(conf: GinConfig= {}) {
    let c: GinConfig = { ...this.defaultConfig,
      ...conf,
    };

    if (conf.request) {
      c.request = {
        ...this.defaultConfig.request,
        ...conf.request
      };
    }

    if (conf.sites) {
      c.sites = {
        ...this.defaultConfig.sites,
        ...conf.sites
      };
    }

    return c;
  }
}

export const ginConfig = new Config();
export default ginConfig;




