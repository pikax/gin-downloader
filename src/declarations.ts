/**
 * Created by rodriguesc on 24/03/2017.
 */

import {HTMLDocument} from "libxmljs";
import {URL} from "url";

export interface MangaXDoc extends HTMLDocument {
  baseUrl: string;
  location: string;
}

export interface Request {
  getHtml(requestedPath: string | URL, params?: any): Promise<string>;
  getBytes(requestedPath: string | URL, params?: any): Promise<any>;
  getDoc(requestedPath: string): Promise<MangaXDoc>;

  postHtml(requestedPath: string | URL, params?: any): Promise<string>;
  postBytes(requestedPath: string | URL, params?: any): Promise<any>;
  postDoc(requestedPath: string| URL, params?: any): Promise<MangaXDoc>;
}

export interface NameHelper {
  toName(name: string): string;
  resolveUrl(name: string): string;
}


export interface SiteConfig {
  site: string;
  name: string;
  mangas_url: string;
  latest_url: string;

}

export interface MangaSource {
  name: string;
  src: string;
}

export interface Chapter {
  number?: number;
  src: string;
  volume?: string;
  name?: string;
}


export interface MangaInfo {
  image: string;
  title: string;

  synonyms?: string[];
  released?: string;
  authors?: string[];
  artists?: string[];
  genres?: string[];
  synopsis?: string;

  status?: string;
  ranked?: string;
  rating?: string;
  scanlators?: string[];
  similarmanga?: string[];
  direction?: string;

  views?: string;
}


export interface ImageSource {
  src: string;
  name: string;
}

export interface SiteParser {
  mangas(doc: MangaXDoc): Promise<MangaSource[]> | MangaSource[];
  latest(doc: MangaXDoc): Promise<Chapter[]> | Chapter[];

  info(doc: MangaXDoc): Promise<MangaInfo> | MangaInfo;
  chapters(doc: MangaXDoc): Promise<Chapter[]> | Chapter[];

  imagesPaths(doc: MangaXDoc): string[];
  image(html: string): string;
}

export interface Site {
  mangas(filter?: FilterSupport): Promise<MangaSource[]>;
  latest(): Promise<Chapter[]>;

  info(name: string): Promise<MangaInfo>;
  chapters(name: string): Promise<Chapter[]>;

  infoChapters(name: string): Promise<{info: MangaInfo, chapters: Chapter[]}>;

  images(name: string, chapter: number): Promise<Promise<ImageSource>[]>;
}



declare global {
  interface Promise<T>{
    tap<U>(result?: (result: any) => U): Promise<T>;
  }

  interface String {
    lastDigit(): number;
    firstDigit(): number;
    leftTrim(): string;
    decodeEscapeSequence(): string;
    getMatches(regex: RegExp, index: number): string[];
  }
}



Promise.prototype.tap = function <T>(func: (arg: T) => void) {
  return this.then((x: T) => {
    func(x);
    return x;
  });
};



const regexLastDigit = /\d+(\.\d{1,2})?$/g;
const regexFirstDigit = /\d+(\.\d{1,2})?/g;

String.prototype.lastDigit = function(){
  let match = this.match(regexLastDigit);
  return (match && +match[0]) || null;
};

String.prototype.firstDigit = function(){
  let match = this.match(regexFirstDigit);
  return (match && +match[0]) || null;
};

String.prototype.leftTrim = function() {
  return this.replace(/^\s+/, "");
};

String.prototype.decodeEscapeSequence = function() {
  return this.replace(/\\x([0-9A-Fa-f]{2})/g, function() {
    return String.fromCharCode(parseInt(arguments[1], 16));
  });
};


String.prototype.getMatches = function(regex: RegExp, index: number) {
  index || (index = 1); // default to the first capturing group
  let matches = [];
  let match;
  while (match = regex.exec(this)) {
    matches.push(match[index]);
  }
  return matches;
};





export enum Genre {
  Action = <any>"Action",
  Adult = <any>"Adult",
  Adventure = <any>"Adventure",
  AwardWinning = <any>"Award Winning",
  Comedy = <any>"Comedy",
  Comic = <any>"Comic",
  Cooking = <any>"Cooking",
  Doujinshi = <any>"Doujinshi",
  Drama = <any>"Drama",
  Ecchi = <any>"Ecchi",
  Fantasy = <any>"Fantasy",
  FourKoma = <any>"4-Koma",
  GenderBender = <any>"Gender Bender",
  Harem = <any>"Harem",
  Historical = <any>"Historical",
  Horror = <any>"Horror",
  Josei = <any>"Josei",
  Lolicon = <any>"Lolicon",
  Manga = <any>"Manga",
  Manhua = <any>"Manhua",
  Manhwa = <any>"Manhwa",
  MartialArts = <any>"Martial Arts",
  Mature = <any>"Mature",
  Mecha = <any>"Mecha",
  Medical = <any>"Medical",
  Music = <any>"Music",
  Mystery = <any>"Mystery",
  Oneshot = <any>"Oneshot",
  Psychological = <any>"Psychological",
  Romance = <any>"Romance",
  SchoolLife = <any>"School Life",
  SciFi = <any>"Sci-fi",
  Seinen = <any>"Seinen",
  Shotacon = <any>"Shotacon",
  Shoujo = <any>"Shoujo",
  ShoujoAi = <any>"Shoujo Ai",
  Shounen = <any>"Shounen",
  ShounenAi = <any>"Shounen Ai",
  SliceOfLife = <any>"Slice of Life",
  Smut = <any>"Smut",
  Sports = <any>"Sports",
  Supernatural = <any>"Supernatural",
  Tragedy = <any>"Tragedy",
  Webtoon = <any>"Webtoon",
  Yaoi = <any>"Yaoi",
  Yuri = <any>"Yuri",

  NoChapters = <any>"[no chapters]"
}


export enum FilterCondition {
  Contains,
  NotContains,
  StartsWith,
  EndsWith,
  Less,
  Greater,
  LessThan,
  GreaterThan
}


export interface FilterSupport {
  name?: string;
  page?: number;
  search?: {
    name: {name: string, condition: FilterCondition}
  };
  genres?: Genre[];
  outGenres?: Genre[];
}

