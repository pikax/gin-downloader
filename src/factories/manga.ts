import {
    ChapterSource,
    IChapter,
    IChapterExtended, IChapterResolver,
    ILazy, ImageCollection,
    IManga,
    IMangaExtended,
    IMangaFactory, IMangaInfoResolver, IMangaObject, Info,
    IReadOnlyManga,
    MangaInfo
} from "../interface";
import {IMangaConfig, IMangaParser, IMangaRequest, IMangaRequestFactory, MangaRequestResult} from "../manga/interface";
import {Lazy} from "../util/lazy";

export class ExtendedMangaFactory implements IMangaFactory {

    constructor(private _requestFactory: IMangaRequestFactory, private _parser: IMangaParser) {

    }


    create(src: string, manga: IManga): IMangaObject {
        const lazySrc = this.lazyRequest(src);

        const lazyChapters = new Lazy(() => this.resolveChaptersPromise(lazySrc));
        const lazyInfo = new Lazy(() => this.resolveInfoPromise(lazySrc));


    }


    private lazyRequest(src: string) {
        return new Lazy(() => this._requestFactory.request({uri: src}));
    }


    private async resolveChaptersPromise(lazy: ILazy<Promise<IMangaRequest>>) {
        const result = await lazy.value;

        return this._parser.chapters(result);
    }

    private async resolveInfoPromise(lazy: ILazy<Promise<IMangaRequest>>) {
        const result = await lazy.value;

        return this._parser.info(result);
    }


}


interface IImageResolver {
}

export class MangaObject implements IMangaObject, IReadOnlyManga {
    private _chapters: ChapterSource[];
    private _info: Promise<Info>;

    get name() {
        return this._manga.name;
    }

    get status() {
        return this._manga.status;
    }

    get mature() {
        return this._manga.mature;
    }

    get image() {
        return this._manga.image;
    }


    constructor(private _src: string, private _manga: IManga, private _chapterResolver: IChapterResolver, private _infoResolver: IMangaInfoResolver, private _imagesResolver: IImageResolver) {
    }

    async chapters(): Promise<IChapter[]> {
        const chapters = await this._chapterResolver.chapters(this._src);
        this._chapters = chapters;

        return chapters.map(x => ({...x, src: undefined}));
    }

    async images(chapter: IChapter): Promise<ImageCollection> {
        const {chap_number} = chapter;

        // resolve chapters if not resolved yet
        if (!this._chapters) {
            await this.chapters();
        }

        const chapters = this._chapters;

        const chap = chapters.find(x => x.chap_number === chap_number);
        if (!chap) {
            throw new Error("Chapter not found");
        }

        const src = chap.src;
        return this._imagesResolver.images(src);
    }

    info(): Promise<Info> {
        return this._info || (this._info = this._infoResolver.info(this._src));
    }
}



