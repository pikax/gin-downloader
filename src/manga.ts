import {
    ChapterSource, IChapter,
    IChapterResolver, IChapterResolverDependency, IImageResolver, IImageResolverDependency, IInfoResolverDependency,
    ImageCollection,
    IManga,
    IMangaInfoResolver,
    IMangaObject, IReadOnlyManga,
    MangaInfo
} from "./interface";


export type MangaObjectDependencies = IImageResolverDependency & IChapterResolverDependency & IInfoResolverDependency;

export class MangaObject implements IMangaObject, IReadOnlyManga {
    private _chapters: ChapterSource[];
    private _info: Promise<MangaInfo>;


    private _chapterResolver: IChapterResolver;
    private _infoResolver: IMangaInfoResolver;
    private _imageResolver: IImageResolver;

    get name() {
        return this._manga.name;
    }

    get status() {
        return this._manga.status;
    }

    get mature() {
        return this._manga.mature;
    }

    get image() {
        return this._manga.image;
    }


    manga() {
        return {...this._manga};
    }

    constructor(dependencies: MangaObjectDependencies, private _src: string, private _manga: IManga) {
        this._chapterResolver = dependencies.chapterResolver;
        this._infoResolver = dependencies.infoResolver;
        this._imageResolver = dependencies.imageResolver;
    }

    async chapters(): Promise<IChapter[]> {
        const chapters = await this._chapterResolver.chapters(this._src);
        this._chapters = chapters;

        return chapters.map(x => ({
            chap_number: x.chap_number,
            volume: x.volume,
            name: x.name,

            language: x.language,
            scanlator: x.scanlator,
            dateAdded: x.dateAdded,

            licensed: x.licensed
        }));
    }

    async images(chapter: IChapter | string | number): Promise<ImageCollection> {
        const chap_number = typeof chapter === "string"
            ? chapter
            : typeof chapter === "number"
                ? chapter.toString()
                : !!chapter.chap_number
                    ? chapter.chap_number
                    : undefined;

        if (!chap_number) {
            throw new Error("Invalid chapter provider");
        }


        // resolve chapters if not resolved yet
        if (!this._chapters) {
            await this.chapters();
        }

        const chapters = this._chapters;

        const chap = chapters.find(x => x.chap_number === chap_number);
        if (!chap) {
            throw new Error("Chapter not found");
        }

        const src = chap.src;
        return this._imageResolver.images(src);
    }

    info(): Promise<MangaInfo> {
        return this._info || (this._info = this._infoResolver.info(this._src));
    }
}
