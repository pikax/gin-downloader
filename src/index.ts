/**
 * Created by david on 25/03/2017.
 */


import "./declarations";

import {MangaFox} from "./sites/mangafox";
import {MangaHere} from "./sites/mangahere";
import {MangaPanda} from "./sites/mangapanda";


export const gin = {
  MangaFox,
  MangaHere,
  MangaPanda
};

export default gin;

