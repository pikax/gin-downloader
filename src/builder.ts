import {MangaHere, MangaHereDependencies} from "./manga/mangahere";
import {MangaHereBuilder} from "./manga/mangahere/builder";
import {RequestRetryStrategy} from "./strategies/retry_strategy";
import {ConcurrentQueueRequestFactory} from "./factories/concurrentQueueRequest";
import {DefaultResolverFactory} from "./factories/resolvers";

const Queue = require("promise-queue");

export type GinBuilderDependencies = {};

export class GinBuilder {
    private _mangahere: MangaHere;
    private _mangaHereBuilder: MangaHereBuilder;
    private _requestFactory: ConcurrentQueueRequestFactory;
    private _requestStrategy: RequestRetryStrategy;
    private _resolverFactory: DefaultResolverFactory;


    get mangahere() {
        return this._mangahere || (this._mangahere = this.buildMangaHere());
    }


    constructor(dependencies: GinBuilderDependencies = {}) {
        this._requestStrategy = new RequestRetryStrategy();
        this._resolverFactory = new DefaultResolverFactory();
        this._requestFactory = new ConcurrentQueueRequestFactory(this._requestStrategy, new Queue(3));

        this._mangaHereBuilder = new MangaHereBuilder();
    }


    private buildMangaHere() {
        return new MangaHere({requestFactory: this._requestFactory, resolverFactory: this._resolverFactory});
    }


}

