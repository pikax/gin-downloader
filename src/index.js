/**
 * Created by david on 19/03/2017.
 */

import fs from 'fs';

import builder from './sites';


let sites = {};

const sitesDir = __dirname + '/sites/';
fs.readdirSync(sitesDir).forEach((file)=> {
  let fp = sitesDir + file;

  if( fs.statSync(fp).isFile())
    return; //nothing to do here

  let site = require(fp).default;
  let parser = require(fp + '/parser').default;
  let config = require(fp + '/config').default;


  let def = new builder(site,config,parser);

  let s = site;

  s.mangas = s.mangas || def.mangas.bind(def);

  site[config.name] = s;
});


modules.export = site;


