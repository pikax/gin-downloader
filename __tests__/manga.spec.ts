import {MangaObject, MangaObjectDependencies} from "../src/manga";
import {IChapterResolver, IManga} from "../src";


describe("Manga Object", () => {
    const chapterResolver: any = {
        chapters: (src) => Promise.resolve([{chap_number: "1", name: "test"}])
    };

    const imageResolver: any = {
        images: (src) => Promise.resolve([{value: Promise.resolve({src: "ImageSrc", name: "test"})}])
    };


    const infoResolver: any = {
        info: (src) => Promise.resolve([{image: "test", title: "test"}])
    };


    const dependencies: MangaObjectDependencies = {
        chapterResolver,
        imageResolver,
        infoResolver,
    };

    const mockedManga: IManga = {
        name: "testName",
        image: "testImage",
        mature: false,
        status: "O"
    };


    it("should get values from passed manga", () => {
        const src = "testSrc";

        const manga = new MangaObject(dependencies, src, mockedManga);

        expect(manga.name).toBe(mockedManga.name);
        expect(manga.image).toBe(mockedManga.image);
        expect(manga.mature).toBe(mockedManga.mature);
        expect(manga.status).toBe(mockedManga.status);

    });

    it("should get manga from passed manga", () => {
        const src = "testSrc";

        const manga = new MangaObject(dependencies, src, mockedManga);


        const result = manga.manga();

        expect(result).not.toBe(mockedManga); // it should return a different object
        expect(result).toEqual(mockedManga);
    });


    it("should get chapters", async () => {
        const src = "testSrc";

        const expected = await chapterResolver.chapters(src);

        const manga = new MangaObject(dependencies, src, mockedManga);

        const chapters = await manga.chapters();

        expect(chapters).toEqual(expected);
    });

    it("should get images from passed chapter", async () => {
        const src = "testSrc";

        const expected = await imageResolver.images(src).then(x => x[0].value);

        const manga = new MangaObject(dependencies, src, mockedManga);

        const chapters = await manga.chapters();
        const images = await manga.images(chapters[0]).then(x => x[0].value);


        expect(images).toEqual(expected);
    });

    it("should get images from chapter number", async () => {
        const src = "testSrc";

        const expected = await imageResolver.images(src).then(x => x[0].value);

        const manga = new MangaObject(dependencies, src, mockedManga);

        const images = await manga.images(1).then(x => x[0].value);


        expect(images).toEqual(expected);

    });

    it("should get images from chapter string", async () => {
        const src = "testSrc";

        const expected = await imageResolver.images(src).then(x => x[0].value);

        const manga = new MangaObject(dependencies, src, mockedManga);

        const images = await manga.images("1").then(x => x[0].value);


        expect(images).toEqual(expected);

    });

    it("should throw error if chapter not found", () => {
        const src = "testSrc";

        const manga = new MangaObject(dependencies, src, mockedManga);

        expect(manga.images(-1)).rejects.toThrowErrorMatchingSnapshot();
    });

    it("should throw error if invalid param on chapter", () => {
        const src = "testSrc";
        const manga = new MangaObject(dependencies, src, mockedManga);

        expect(manga.images({invalid: true} as any)).rejects.toThrowErrorMatchingSnapshot();
    });


    it("should get info", async () => {
        const src = "testSrc";

        const expected = await infoResolver.info(src);
        const manga = new MangaObject(dependencies, src, mockedManga);

        const result = await manga.info();

        expect(result).toEqual(expected);
    });


});

