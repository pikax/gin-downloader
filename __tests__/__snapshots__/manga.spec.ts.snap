// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Manga Object should throw error if chapter not found 1`] = `"Chapter not found"`;

exports[`Manga Object should throw error if invalid param on chapter 1`] = `"Invalid chapter provider"`;
